File: server/repositories/alertsRepo.mjs
import { readSeedJson } from '../utils/seedLoader.mjs';
import { runOracleQuery } from '../db/oracleClient.mjs';
import { getServerConfig } from '../config/env.mjs';
import { createOracleUnavailableError } from '../errors/oracleErrors.mjs';
import { recordApiFallbackMetric } from '../observability/metricsStore.mjs';
import { readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ALERT_READS_FILE = path.join(__dirname, '..', 'data', 'alerts_reads_state.json');

function formatRelative(dateValue) {
  const dt = new Date(dateValue);
  if (Number.isNaN(dt.getTime())) return 'recentemente';
  const diffHours = Math.max(0, Math.floor((Date.now() - dt.getTime()) / 3600000));
  if (diffHours < 1) return 'agora';
  if (diffHours < 24) return `${diffHours} h atras`;
  return `${Math.floor(diffHours / 24)} d atras`;
}

function mapSeverity(level) {
  const g = String(level || '').toLowerCase();
  if (g === 'alta' || g === 'critica') return 'critical';
  if (g === 'media') return 'warning';
  return 'info';
}

function normalizeAlertText(value, fallback) {
  if (value === null || value === undefined) {
    return fallback;
  }

  if (typeof value === 'string') {
    const raw = value.trim();
    if (!raw || raw === '[object Object]') {
      return fallback;
    }

    if (raw.startsWith('{') || raw.startsWith('[')) {
      try {
        const parsed = JSON.parse(raw);
        if (parsed && typeof parsed === 'object') {
          const candidate = parsed.description || parsed.descricao || parsed.message || parsed.mensagem || parsed.title || parsed.titulo;
          if (candidate) {
            return String(candidate).slice(0, 240);
          }
        }
      } catch {}
    }

    return raw.slice(0, 240);
  }

  if (typeof value === 'object') {
    const candidate = value.description || value.descricao || value.message || value.mensagem || value.title || value.titulo;
    if (candidate) {
      return String(candidate).slice(0, 240);
    }
    return fallback;
  }

  const scalar = String(value).trim();
  if (!scalar || scalar === '[object Object]') {
    return fallback;
  }
  return scalar.slice(0, 240);
}

async function readReadsState() {
  try {
    const raw = await readFile(ALERT_READS_FILE, 'utf-8');
    const parsed = JSON.parse(raw);
    if (parsed && typeof parsed === 'object') {
      return parsed;
    }
    return {};
  } catch (error) {
    if (String(error?.code || '') === 'ENOENT') {
      return {};
    }
    return {};
  }
}

async function writeReadsState(state) {
  await writeFile(ALERT_READS_FILE, JSON.stringify(state, null, 2), 'utf-8');
}

function applyReadState(items, condominiumId, readsState) {
  const tenantKey = String(condominiumId);
  const tenantState = readsState?.[tenantKey] || {};

  return items.map((item) => {
    const readState = tenantState[String(item.id)] || null;
    const read = Boolean(readState?.read);
    const readAt = readState?.readAt || null;
    const readBy = readState?.readBy || null;
    return {
      ...item,
      status: read ? 'read' : 'active',
      read,
      readAt,
      readBy,
    };
  });
}

export async function getAlertsData(condominiumId = 1) {
  const { dbDialect, allowOracleSeedFallback } = getServerConfig();
  const readsState = await readReadsState();

  if (dbDialect === 'oracle') {
    try {
      const rows = await runOracleQuery(`
        select
          alert_id,
          condominio_id,
          data_detectada,
          tipo_anomalia,
          descricao_anomalia,
          gravidade
        from mart.vw_alerts_operational
        where condominio_id = :condominiumId
        order by data_detectada desc
        fetch first 50 rows only
      `, { condominiumId });

      if (rows) {
        const baseItems = rows.map((row) => ({
          id: String(row.ALERT_ID),
          condominiumId: Number(row.CONDOMINIO_ID || 0) || null,
          severity: mapSeverity(row.GRAVIDADE),
          title: normalizeAlertText(row.TIPO_ANOMALIA, 'anomalia detectada').replaceAll('_', ' '),
          description: normalizeAlertText(row.DESCRICAO_ANOMALIA, 'Anomalia detectada automaticamente'),
          time: formatRelative(row.DATA_DETECTADA),
        }));
        const items = applyReadState(baseItems, condominiumId, readsState);

        return { activeCount: items.filter((item) => item.status === 'active').length, items };
      }
    } catch (error) {
      if (!allowOracleSeedFallback) {
        throw createOracleUnavailableError(error);
      }
      recordApiFallbackMetric('alerts', 'oracle_fallback_seed');
    }
  }

  const seed = readSeedJson('alerts.json');
  const baseItems = seed.items
    .map((item) => ({ ...item, condominiumId: 1 }))
    .filter((item) => item.condominiumId === condominiumId);
  const items = applyReadState(baseItems, condominiumId, readsState);
  return { activeCount: items.filter((item) => item.status === 'active').length, items };
}

export async function markAlertAsRead(condominiumId = 1, alertId, actorSub = null) {
  const payload = await getAlertsData(condominiumId);
  const exists = payload.items.some((item) => String(item.id) === String(alertId));
  if (!exists) {
    return null;
  }

  const state = await readReadsState();
  const tenantKey = String(condominiumId);
  state[tenantKey] = state[tenantKey] || {};
  state[tenantKey][String(alertId)] = {
    read: true,
    readAt: new Date().toISOString(),
    readBy: actorSub || null,
  };
  await writeReadsState(state);

  const refreshed = await getAlertsData(condominiumId);
  return refreshed.items.find((item) => String(item.id) === String(alertId)) || null;
}

File: backend/app/repositories/alerts_repo.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from app.core.config import settings
from app.core.errors import create_oracle_unavailable_error
from app.db.oracle_client import run_oracle_query
from app.observability.metrics_store import record_api_fallback_metric
from app.repositories.state_store import read_json_state, write_json_state
from app.utils.seed_loader import read_seed_json

READS_FILE = Path(__file__).resolve().parents[3] / "backend" / "data" / "alerts_reads_state.json"


def _format_relative(value: Any) -> str:
    try:
        dt = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return "recentemente"
    diff = datetime.now(timezone.utc) - dt.astimezone(timezone.utc)
    hours = max(0, int(diff.total_seconds() // 3600))
    if hours < 1:
        return "agora"
    if hours < 24:
        return f"{hours} h atras"
    return f"{hours // 24} d atras"


def _map_severity(value: Any) -> str:
    raw = str(value or "").lower()
    if raw in {"alta", "critica"}:
        return "critical"
    if raw == "media":
        return "warning"
    return "info"


def _normalize_text(value: Any, fallback: str) -> str:
    if value is None:
        return fallback
    raw = str(value).strip()
    if not raw or raw == "[object Object]":
        return fallback
    return raw[:240]


def _apply_read_state(items: list[dict[str, Any]], condominium_id: int, state: dict[str, Any]) -> list[dict[str, Any]]:
    tenant = state.get(str(condominium_id), {})
    out = []
    for item in items:
        patch = tenant.get(str(item["id"]), {})
        read = bool(patch.get("read"))
        out.append(
            {
                **item,
                "status": "read" if read else "active",
                "read": read,
                "readAt": patch.get("readAt"),
                "readBy": patch.get("readBy"),
            }
        )
    return out


async def get_alerts_data(condominium_id: int = 1) -> dict[str, Any]:
    reads_state = await read_json_state(READS_FILE)

    if settings.db_dialect == "oracle":
        try:
            rows = await run_oracle_query(
                """
                select alert_id, condominio_id, data_detectada, tipo_anomalia, descricao_anomalia, gravidade
                from mart.vw_alerts_operational
                where condominio_id = :condominiumId
                order by data_detectada desc
                fetch first 50 rows only
                """,
                {"condominiumId": condominium_id},
            )
            if rows is not None:
                base_items = [
                    {
                        "id": str(row.get("ALERT_ID")),
                        "condominiumId": int(row.get("CONDOMINIO_ID") or 0) or None,
                        "severity": _map_severity(row.get("GRAVIDADE")),
                        "title": _normalize_text(row.get("TIPO_ANOMALIA"), "anomalia detectada").replace("_", " "),
                        "description": _normalize_text(row.get("DESCRICAO_ANOMALIA"), "Anomalia detectada automaticamente"),
                        "time": _format_relative(row.get("DATA_DETECTADA")),
                    }
                    for row in rows
                ]
                items = _apply_read_state(base_items, condominium_id, reads_state)
                return {"activeCount": len([i for i in items if i["status"] == "active"]), "items": items}
        except Exception as exc:
            if not settings.allow_oracle_seed_fallback:
                raise create_oracle_unavailable_error(exc)
            record_api_fallback_metric("alerts", "oracle_fallback_seed")

    seed = read_seed_json("alerts.json")
    base_items = [{**item, "condominiumId": 1} for item in seed.get("items", [])]
    items = _apply_read_state([i for i in base_items if i.get("condominiumId") == condominium_id], condominium_id, reads_state)
    return {"activeCount": len([i for i in items if i["status"] == "active"]), "items": items}


async def mark_alert_as_read(condominium_id: int, alert_id: str, actor_sub: str | None = None) -> dict[str, Any] | None:
    payload = await get_alerts_data(condominium_id)
    exists = any(str(item["id"]) == str(alert_id) for item in payload["items"])
    if not exists:
        return None

    state = await read_json_state(READS_FILE)
    tenant = state.setdefault(str(condominium_id), {})
    tenant[str(alert_id)] = {"read": True, "readAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"), "readBy": actor_sub}
    await write_json_state(READS_FILE, state)

    refreshed = await get_alerts_data(condominium_id)
    return next((item for item in refreshed["items"] if str(item["id"]) == str(alert_id)), None)
